File: Projects/po_projekt2/World.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package po_projekt2;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.*;
import po_projekt2.worldElements.Board;
import po_projekt2.Organism;
import po_projekt2.worldElements.WorldDrafter;

import po_projekt2.Organism.typeOfOrganism;
import po_projekt2.organisms.animals.Antelope;
import po_projekt2.organisms.animals.CyberSheep;
import po_projekt2.organisms.animals.Fox;
import po_projekt2.organisms.animals.Human;
import po_projekt2.organisms.animals.Sheep;
import po_projekt2.organisms.animals.Turtle;
import po_projekt2.organisms.animals.Wolf;
import po_projekt2.organisms.plants.Belladonna;
import po_projekt2.organisms.plants.Grass;
import po_projekt2.organisms.plants.Guarana;
import po_projekt2.organisms.plants.Hogweed;
import po_projekt2.organisms.plants.Milt;

import po_projekt2.worldElements.Coordinates;

/**
 *
 * @author Bogna
 */
public class World {
    private Board board;
    private Narrator narrator;
    private int numberOfTurn, numberOfOrganisms;
    private boolean isHumanAlive, isEndOfGame, isEndOfTurn;
    private ArrayList<Organism> organisms;
    private WorldDrafter worldDrafter;
    private Human human;
    
    private void sortOrganisms() {
        Collections.sort(organisms, new Comparator<Organism>() {
            @Override
            public int compare(Organism o1, Organism o2) {
                if (o1.getInitiative() != o2.getInitiative())
                    return Integer.valueOf(o2.getInitiative()).compareTo(o1.getInitiative());
                else
                    return Integer.valueOf(o1.getWhenWasBorn()).compareTo(o2.getWhenWasBorn());
            }
        });
    }
    
    public World(int numberOfOrganisms, int sizeX, int sizeY, WorldDrafter worldDrafter){
        this.numberOfTurn = 0;
        this.isHumanAlive = true;
        this.isEndOfGame = false;
        
        organisms = new ArrayList<>();
        board = new Board(sizeX, sizeY);
        this.worldDrafter = worldDrafter;
        
        Coordinates c = board.randEmptyArea();
	Organism tmp = new Human(c, this);
	addOrganism(tmp);
        this.human = (Human)tmp;
        
        for(int i = 1; i < numberOfOrganisms; i++){
            Coordinates coords = board.randEmptyArea();
            Organism.typeOfOrganism type = human.randType();
            Organism o = createNewOrganism(type, coords);
            this.addOrganism(o);
        }
        
        this.isEndOfTurn = true;
    }
    
    public World(WorldDrafter worldDrafter) {
        this.numberOfTurn = 0;
        this.isHumanAlive = true;
        this.isEndOfGame = false;
        
        organisms = new ArrayList<>();
        this.worldDrafter = worldDrafter;
        this.isEndOfTurn = true;
    }
    
    
    private static class Narrator{
        private static String comments;
        
        public Narrator(){
            comments = "";
        }
        
        public static void addComment(String comment){
            comments += comment;
        }
        
        public static String getComments(){
            return comments;
        }
        
        public static void clearComments(){
            comments = "";
        }
    }
    
    
    public boolean saveWorld(String fileName){
        try{
            File file = new File(fileName);
            file.createNewFile();
            
            PrintWriter pw = new PrintWriter(file);
            pw.write(board.getSizeX() + " " + board.getSizeY() + " " + organisms.size() + " " + numberOfTurn + " " + isHumanAlive + "\n");
            for (int i = 0; i < organisms.size(); i++) {
		pw.write(organisms.get(i).getType() + " " + organisms.get(i).getStrenght() + " " + organisms.get(i).getInitiative() + " " + organisms.get(i).getWhenWasBorn() + " " + organisms.get(i).getCoords().getX() + " " + organisms.get(i).getCoords().getY());
		if (organisms.get(i) instanceof Human) {
                    Human tmp = (Human)organisms.get(i);
                    pw.write(" " + tmp.getSkillIsActive() + " " + tmp.getNumberOfTurnsSkillWillBeActive());
		}
		pw.write("\n");
            }
            pw.close();
            return true;
        } catch (IOException e) {
            System.out.println("Error: " + e);
            return false;
        }
    }
    
    
    public boolean loadWorld(String fileName){
        try{
            File file = new File(fileName);

            Scanner scan = new Scanner(file);
            String line = scan.nextLine();
            String[] elements = line.split(" ");
            int sizeX = Integer.parseInt(elements[0]);
            int sizeY = Integer.parseInt(elements[1]);
            
            board = new Board(sizeX, sizeY);
            this.numberOfOrganisms = Integer.parseInt(elements[2]);
            this.numberOfTurn = Integer.parseInt(elements[3]);
            this.isHumanAlive = Boolean.parseBoolean(elements[4]);
            while (scan.hasNextLine()) {
                line = scan.nextLine();
                elements = line.split(" ");
                Organism.typeOfOrganism type = Organism.typeOfOrganism.valueOf(elements[0]);
                int x = Integer.parseInt(elements[4]);
                int y = Integer.parseInt(elements[5]);
                Coordinates coords = new Coordinates(x, y);
                Organism o = null;
                int strenght = Integer.parseInt(elements[1]);
                int whenWasBorn = Integer.parseInt(elements[3]);
                
                if(type == Organism.typeOfOrganism.HUMAN){
                    o = new Human(coords, this);
                    o.setStrenght(strenght);
                    human = (Human)o;
                    boolean skillIsActive = Boolean.parseBoolean(elements[4]);
                    int numberOfTurnWillBeActive = Integer.parseInt(elements[3]);
                    human.setSkillIsActive(skillIsActive);
                    human.setNumberOfTurnsSkillWillBeActive(numberOfTurnWillBeActive);
                }else{
                    o = createNewOrganism(type, coords);
                    o.setStrenght(strenght);
                    o.setWhenWasBorn(whenWasBorn);
                }
                this.addOrganism(o);
            }
            return true;
        }catch (
                IOException e) {
            System.out.println("Error: " + e);
            return false;
        }
    }
    
    
    public void makeTurn() {
        narrator.clearComments();
	if (isEndOfGame) {
            return;
	}
        this.sortOrganisms();
	for (int i = 0; i < organisms.size(); i++) {
            if(organisms.get(i).getWhenWasBorn() != numberOfTurn){
                organisms.get(i).action();
            }
	}
	numberOfTurn++;
    }
    
    public void addComment(String comment){
        narrator.addComment(comment);
    }
    
    public void clearComment(){
        narrator.clearComments();
    }
    
    public Board getBoard(){
        return this.board;
    }
    
    public String getComments(){
        return narrator.getComments();
    }
    
    public WorldDrafter getWorldDrafter(){
        return this.worldDrafter;
    }
    
    public int getNumberOfTurn(){
        return this.numberOfTurn;
    }
    
    public int getNumberOfAnimals(){
        return this.numberOfOrganisms;
    }
    
    public boolean getIsHumanAlive(){
        return this.isHumanAlive;
    }
    
    public boolean getIsEndOfGame(){
        return this.isEndOfGame;
    }
    
    public boolean getIsEndOfTurn(){
        return this.isEndOfTurn;
    }
    
    public Human getHuman(){
        return this.human;
    }
    
    public void setIsHumanAlive(boolean isHumanAlive){
        this.isHumanAlive = isHumanAlive;
    }
    
    public void setIsEndOfGame(boolean isEndOfGame){
        this.isHumanAlive = isHumanAlive;
    }
    
    public void setIsEndOfTurn(boolean isEndOfTurn){
        this.isEndOfTurn = isEndOfTurn;
    }
    
    public void addOrganism(Organism newOrganism){
        organisms.add(newOrganism);
        this.numberOfOrganisms = organisms.size();
        board.setOrganism(newOrganism);
    }
    
    public Organism createNewOrganism(typeOfOrganism type, Coordinates newCoords){
        Organism newOrganism;
        if(type == Organism.typeOfOrganism.WOLF){
            newOrganism = new Wolf(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.SHEEP){
            newOrganism = new Sheep(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.FOX){
            newOrganism = new Fox(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.TURTLE){
            newOrganism = new Turtle(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.ANTELOPE){
            newOrganism = new Antelope(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.CYBER_SHEEP){
            newOrganism = new CyberSheep(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.GRASS){
            newOrganism = new Grass(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.MILT){
            newOrganism = new Milt(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.GUARANA){
            newOrganism = new Guarana(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.BELLADONNA){
            newOrganism = new Belladonna(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        } else if(type == Organism.typeOfOrganism.HOGWEED){
            newOrganism = new Hogweed(this.getNumberOfTurn(), newCoords, this);
            return newOrganism;
        }
        return null;
    }
    
    public void deleteOrganism(Organism o){
        int i = 0;
        while(o != organisms.get(i)){
            i++;
        }
        organisms.remove(i);
        this.numberOfOrganisms = organisms.size();
    }
    
    
}

File: Project_1-c++/World.cpp
#include "World.h"

#include "Human.h"
#include "Wolf.h"
#include "Turtle.h"
#include "Sheep.h"
#include "Fox.h"
#include "Antelope.h"
#include "Board.h"

#include "Grass.h"
#include "Milt.h"
#include "Guarana.h"
#include "Belladonna.h"
#include "Hogweed.h"

#include <algorithm>
#include <iostream>
#include <fstream>
#include <conio.h>
using namespace std;

#define MAX_DIRECTIONS 4

enum inputs {
	if_arrows = 224,	//sterowanie
	arrow_up = 72,
	arrow_down = 80,
	arrow_left = 75,
	arrow_right = 77,
	space = 32,		//aktywacja umiejêtnoœci
	esc = 27,		//zamykanie gry
	enter = 13,		//pzechodzenie do kolejnej tury
	s = 115,		//ma³e s  <- do zapisywania
	S = 83			//du¿e s
};

World::World(int sizeX, int sizeY) {
	numberOfOrganisms = 0;
	board = new Board(sizeX, sizeY);
	turn = 0;
	isHumanAlive = true;
	organisms.reserve(0);
}

World::World(int numberOfOrganisms, int sizeX, int sizeY) {
	this->numberOfOrganisms = numberOfOrganisms;
	board = new Board(sizeX, sizeY);
	Coordinates c = board->randEmptyArea();
	Organism* human = new Human(c, this);
	this->human = dynamic_cast<Human*>(human);
	addOrganism(human);
	for (int k = 0; k < numberOfOrganisms - 1; k++) {
		Coordinates coords = board->randEmptyArea();
		Organism::TypeOfOrganism type = human->randType();
		Organism* o = createNewOrganism((int)type, coords);
		addOrganism(o);
	}
	isHumanAlive = true;
	isEndOfGame = false;
}

World::World(World&& other)
{
	this->turn = other.turn;
	this->board = other.board;
	this->isHumanAlive = other.isHumanAlive;
	this->isEndOfGame = other.isEndOfGame;
	this->organisms = other.organisms;
	this->human = other.human;

	for (size_t i = 0; i < organisms.size(); i++) {
		organisms[i]->SetWorld(this);
	}

	other.turn = 0;
	other.board = nullptr;
	other.isHumanAlive = false;
	other.isEndOfGame = false;
	other.human = nullptr;
}

int World::GetTurn() {
	return turn;
}

Board* World::GetBoard() {
	return board;
}

bool World::GetIsHumanAlive() {
	return isHumanAlive;
}

bool World::GetIsEndOfGame() {
	return isEndOfGame;
}

void World::SetIsHumanAlive(bool isAlive) {
	this->isHumanAlive = isAlive;
}

void World::SetIsEndOfGame(bool isEndOfGame) {
	this->isEndOfGame = isEndOfGame;
}

void World::SetTurn(int turn) {
	this->turn = turn;
}

void World::SetHuman(Human* human) {
	this->human = human;
}

void World::SetNumberOfOrganisms(int number) {
	this->numberOfOrganisms = number;
}

World* World::loadWorld() {
	fstream file;
	string fileName;
	cout << "Loading world from text file.\n";
	bool isFileGood = false;
	cout << "Give file name (add .txt at the end): ";
	while (!isFileGood) {
		cin >> fileName;
		file.open(fileName, fstream::in);
		if (file.good()) {
			cout << "Opening file...\n";
			isFileGood = true;
		}
		else {
			cout << "Wrong file. Try again.\n";
		}
	}
	if (file.is_open()) {
		int x, y, numberOfOrganisms, turn, isHumanAlive;
		file >> x >> y >> numberOfOrganisms >> turn >> isHumanAlive;
		World* tmpWorld = new World(x, y);
		tmpWorld->SetNumberOfOrganisms(numberOfOrganisms);
		tmpWorld->SetIsHumanAlive(isHumanAlive);
		tmpWorld->SetTurn(turn);
		int type, strenght, initiative, whenWasBorn, skillIsActive, numberOfTurnsSkillWillBeActive;
		Organism* tmpOrganism = nullptr;
		for (int i = 0; i < numberOfOrganisms; i++) {
			file >> type >> strenght >> initiative >> whenWasBorn >> x >> y;
			if (type == (int)Organism::TypeOfOrganism::HUMAN) {
				file >> skillIsActive >> numberOfTurnsSkillWillBeActive;
				tmpOrganism = new Human(Coordinates(x, y), tmpWorld);
				tmpOrganism->SetStrenght(strenght);
				Human* human = dynamic_cast<Human*>(tmpOrganism);
				human->SetSkillIsActive(skillIsActive);
				human->SetNumberOfTurnsSkillWillBeActive(numberOfTurnsSkillWillBeActive);
				tmpWorld->SetHuman(human);
			}else {
				tmpOrganism = tmpWorld->createNewOrganism(type, Coordinates(x, y));
				tmpOrganism->SetStrenght(strenght);
				tmpOrganism->SetWhenWasBorn(whenWasBorn);
			}
			tmpWorld->addOrganism(tmpOrganism);
		}
		return tmpWorld;
	}
	file.close();
	return nullptr;
}

void World::saveWorld() {
	fstream file;
	string fileName;
	cout << "Saving world to text file.\n";
	bool isFileGood = false;
	cout << "Give file name (add .txt at the end): ";
	while (!isFileGood) {
		cin >> fileName;
		file.open(fileName, fstream::out);
		if (file.good()) {
			cout << "Opening file...\n";
			isFileGood = true;
		}
		else {
			cout << "Wrong file. Try again.\n";
		}
	}
	file << board->GetSizeX() << " " << board->GetSizeY() << " " << organisms.size() << " " << turn << " " << (int)isHumanAlive << endl;
	for (int i = 0; i < organisms.size(); i++) {
		file << (int)organisms[i]->GetType() << " " << organisms[i]->GetStrenght() << " " << organisms[i]->GetInitiative() << " " << organisms[i]->GetWhenWasBorn() << " " << organisms[i]->GetCoordinates().GetX() << " " << organisms[i]->GetCoordinates().GetY();
		if (dynamic_cast<Human*>(organisms[i])) {
			Human* tmp = dynamic_cast<Human*>(organisms[i]);
			file << " " << tmp->GetSkillIsActive() << " " << tmp->GetNumberOfTurnsSkillWillBeActive();
		}
		file << endl;
	}
	file.close();
	cout << "World saved.\n";

}

void World::drawWorld() {
	board->drawBoard();
	
	cout << endl;
}

void World::printMessage(Organism* winning, Organism* other, bool ifAttack, bool specialAction) {
	if (ifAttack && !specialAction) {
		winning->GetOrganismName();
		cout << " [ x = " << winning->GetCoordinates().GetX() << " ; y = " << winning->GetCoordinates().GetY();
		if (other->GetIsAnimal()){
			cout << " ] kills ";
		}else {
			cout << " ] eats ";
		}
		other->GetOrganismName();
		cout << " [ x = " << other->GetCoordinates().GetX() << " ; y = " << other->GetCoordinates().GetY() << " ]\n";
	}
	else if (specialAction && dynamic_cast<Turtle*>(winning)) {
		winning->GetOrganismName();
		cout << " [ x = " << winning->GetCoordinates().GetX() << " ; y = " << winning->GetCoordinates().GetY() << " ] expels ";
		other->GetOrganismName();
		cout << " [ x = " << other->GetCoordinates().GetX() << " ; y = " << other->GetCoordinates().GetY() << " ]\n";
	}
	else if (specialAction && dynamic_cast<Antelope*>(winning)) {
		winning->GetOrganismName();
		cout << " [ x = " << winning->GetCoordinates().GetX() << " ; y = " << winning->GetCoordinates().GetY() << " ] escapes from ";
		other->GetOrganismName();
		cout << " [ x = " << other->GetCoordinates().GetX() << " ; y = " << other->GetCoordinates().GetY() << " ]\n";
	}
	else {
		cout << "New organism was born. ";
		winning->GetOrganismName();
		cout << " [ x = " << winning->GetCoordinates().GetX() << " ; y = " << winning->GetCoordinates().GetY() << " ]\n";
	}
}

void World::makeTurn() {
	input();
	if (isEndOfGame) {
		return;
	}
	sort(organisms.begin(), organisms.end(), [](Organism* first, Organism* second) {
		if (first->GetInitiative() > second->GetInitiative()) return true;
		else if (first->GetInitiative() == second->GetInitiative()) {
			if (first->GetWhenWasBorn() > second->GetWhenWasBorn()) return true;
			else return false;
		}
		else return false;
		});
	system("cls");
	cout << "Bogna Lew 184757\n";
	switch(isHumanAlive) {
	case 0: cout << "Human is dead. Press enter to move to the next turn.\n";
	case 1: cout << "Arrows - controling human; space - activating special skill (\"Antelope's speed\")\n";
	}
	cout << "S - save world; ESC - quit\n\n";
	
	for (int i = 0; i < organisms.size(); i++) {
		organisms.at(i)->action();
	}
	cout << endl;
	drawWorld();
	turn++;
}

void World::input() {
	while (true) {
		int znak = _getch();
		if (GetIsHumanAlive() && znak == if_arrows) {
			znak = _getch();
			if (znak == arrow_up) {
				human->SetDirection(Organism::Direction::UP);
				return;
			}else if (znak == arrow_down) {
				human->SetDirection(Organism::Direction::DOWN);
				return;
			}else if(znak == arrow_left) {
				human->SetDirection(Organism::Direction::LEFT);
				return;
			}else if(znak == arrow_right) {
				human->SetDirection(Organism::Direction::RIGHT);
				return;
			}
		}else if (GetIsHumanAlive() && znak == space) {
			if (human->GetSkillIsActive()) {
				cout << "Special skill is active. It will be deactivated after: " << human->GetNumberOfTurnsSkillWillBeActive() - 5 << " turns.\n";
			}else if (!human->GetSkillIsActive() && human->GetNumberOfTurnsSkillWillBeActive() > 0) {
				cout << "You cannot activate special skill yet. You will be able to do that after: " << human->GetNumberOfTurnsSkillWillBeActive() << "turns.\n";
			}else {
				human->activateSkill();
				human->SetDirection(Organism::Direction::NONE);
				cout << "Special skill activated: \"Antelope's speed\". The skill will be active for 5 turns.\n";
				return;
			}
		}else if (!GetIsHumanAlive() && (znak == if_arrows || znak == space)) {
			cout << "Human is dead. You can no longer control him.\n";
		}else if (znak == enter && !isHumanAlive) {
			human->SetDirection(Organism::Direction::NONE);
			return;
		}
		else if (znak == s || znak == S) {
			saveWorld();
		}
		else if (znak == esc) {
			SetIsEndOfGame(true);
			return;
		}
	}
}

Organism* World::createNewOrganism(int type, Coordinates coords) {
	switch (type) {
	case (int)Organism::TypeOfOrganism::WOLF: return new Wolf(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::SHEEP : return new Sheep(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::FOX : return new Fox(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::TURTLE : return new Turtle(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::ANTELOPE : return new Antelope(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::GRASS: return new Grass(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::MILT: return new Milt(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::GUARANA: return new Guarana(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::BELLADONNA: return new Belladonna(this->GetTurn(), coords, this);
	case (int)Organism::TypeOfOrganism::HOGWEED: return new Hogweed(this->GetTurn(), coords, this);
	}
}

void World::addOrganism(Organism* organism) {
	board->SetOrganism(organism);
	organisms.push_back(organism);
}

void World::deleteOrganism(Organism* organism) {
	int number = 0;
	for (int i = 0; i < organisms.size(); i++) {
		if (organism == organisms.at(i)) {
			number = i;
			break;
		}
	}
	organisms.erase(organisms.begin() + number);
	Organism* tmp = board->GetOrganism(organism->GetCoordinates());
	tmp = nullptr;
}


World::~World() {}